docs/proposals/20231122-automate-aks-features.md

```diff
@@ -0,0 +1,311 @@
+will be linked to the AzureManagedControlPlane through the standard `cluster.x-k8s.io/cluster-name` label.
+
+To modify parameters on the AKS cluster, either the template or non-template ManagedCluster may be updated.
+CAPZ will propagate changes made to a template to instances of that template. Parameters defined on the
+template take precedence over the same parameters on the instances.
+
+The main difference with [Option 1] that enables ClusterClass is that the same ManagedCluster template
+resource can be referenced by multiple AzureManagedControlPlanes, so this new `ManagedClusterTemplateName`
+field can be defined on the AzureManagedControlPlaneClassSpec so a set of AKS parameters defined once in the
+template can be applied to all Clusters built from that ClusterClass.
+
+This method makes all ManagedCluster fields available to define in a template which could lead to
+misconfiguration if certain parameters that must be unique to a cluster are erroneously shared through a
+template. Since those fields cannot be automatically identified and may evolve between AKS API versions, CAPZ
+will not attempt to categorize ASO ManagedCluster fields that way like it does between the fields present and
+omitted from the `AzureClusterClassSpec` type, for example. CAPZ could document a best-effort list of known
+fields which could or should not be defined in template types and will otherwise rely on AKS to provide
+reasonable error messages for misconfigurations.
+
+Like [Option 1], this method keeps particular versions of CAPZ decoupled from particular API versions of ASO
+resources (including allowing preview versions) and opens the door for streamlined adoption of existing AKS
+clusters.
+
+#### Option 3: CAPZ resource defines an entire unstructured ASO resource inline
+
+This method is functionally equivalent to [Option 2] except that the template resource is defined inline
+within the AzureManagedControlPlane:
+
+```go
+type AzureManagedControlPlaneClassSpec struct {
+	...
+	
+	// ManagedClusterTemplate is the ASO ManagedCluster to be used as a template from which new
+	// ManagedClusters will be created.
+	ManagedClusterTemplate map[string]interface{} `json:"managedClusterTemplate"`
+}
+```
+
+One variant of this method could be to using `string` instead of `map[string]interface{}` for the template
+type, though that would make defining patches unwieldy (like for ClusterClass).
+
+Compared to [Option 2], this method loses schema and webhook validation that would be performed by ASO when
+creating a separate ManagedCluster to serve as a template. That validation would still be performed when CAPZ
+creates the ManagedCluster resource, but that would be some time after the AzureManagedControlPlane is created
+and error messages may not be quite as visible.
+
+#### Option 4: CAPZ resource defines an entire typed ASO resource inline
+
+This method is functionally equivalent to [Option 3] except that the template field's type is the exact same
```

Comment:
Can you expand on how we would support ClusterClass for this option?

Reply:
This would be mostly the same as Options 2:

> This method makes all ManagedCluster fields available to define in a template which could lead to misconfiguration if certain parameters that must be unique to a cluster are erroneously shared through a template. Since those fields cannot be automatically identified and may evolve between AKS API versions, CAPZ will not attempt to categorize ASO ManagedCluster fields that way like it does between the fields present and omitted from the `AzureClusterClassSpec` type, for example. CAPZ could document a best-effort list of known fields which could or should not be defined in template types and will otherwise rely on AKS to provide reasonable error messages for misconfigurations.

For this option specifically, One addition I can think of to documenting which fields should not be defined in a template is enforcing that through webhooks, but I still don't think there's a practical way to segment them in the API type definitions. I'm not sure we could do that for the other options where we have less control over the particular ASO API version being used.

Reply:
^ will add that to the doc